File: src-tauri/src/frontmatter.rs
use std::collections::HashSet;

/// Tolerant parser for the tiny frontmatter dialect notras cares about.
///
/// Only `pinned`, `tags` and `title` are interpreted; everything else is
/// ignored (the TypeScript side preserves unknown keys verbatim when
/// rewriting). Kept as a hand-rolled parser so Rust and TS stay in parity over
/// a deliberately tiny format instead of dragging in a full YAML
/// implementation.
///
/// `title` is read-only: notras resolves a title from it but never authors it,
/// so it round-trips as a foreign key through the TypeScript serializer.
#[derive(Debug, Default, PartialEq)]
pub struct Frontmatter {
    pub pinned: bool,
    pub tags: Vec<String>,
    pub title: Option<String>,
}

pub struct Parsed<'a> {
    pub frontmatter: Frontmatter,
    /// The note body with the frontmatter block stripped.
    pub body: &'a str,
}

fn clean_tag(raw: &str) -> Option<String> {
    let tag = raw
        .trim()
        .trim_matches(|c| c == '"' || c == '\'')
        // Separators are dropped, never kept: a tag carrying one would
        // serialize into `tags: [a, b]` and re-parse as two tags.
        .replace([',', '[', ']'], "")
        .trim()
        .to_lowercase();

    if tag.is_empty() {
        None
    } else {
        Some(tag)
    }
}

/// A `title:` value with surrounding quotes stripped. An empty value counts as
/// absent so title resolution falls through to the next source.
fn clean_title(raw: &str) -> Option<String> {
    let title = raw.trim().trim_matches(|c| c == '"' || c == '\'').trim();

    if title.is_empty() {
        None
    } else {
        Some(title.to_string())
    }
}

fn parse_inline_tags(value: &str) -> Vec<String> {
    value
        .trim()
        .trim_start_matches('[')
        .trim_end_matches(']')
        .split(',')
        .filter_map(clean_tag)
        .collect()
}

pub fn parse(content: &str) -> Parsed<'_> {
    let mut parsed = Parsed {
        frontmatter: Frontmatter::default(),
        body: content,
    };

    let Some(rest) = content
        .strip_prefix("---\n")
        .or_else(|| content.strip_prefix("---\r\n"))
    else {
        return parsed;
    };

    // Find the closing delimiter line.
    let mut close = None;
    let mut offset = 0;
    for line in rest.split_inclusive('\n') {
        let trimmed = line.trim_end();
        if trimmed == "---" || trimmed == "..." {
            close = Some((offset, offset + line.len()));
            break;
        }
        offset += line.len();
    }

    let Some((block_end, body_start)) = close else {
        return parsed;
    };

    let block = &rest[..block_end];
    parsed.body = &rest[body_start..];

    let mut in_tags_list = false;
    for line in block.lines() {
        let trimmed = line.trim_end();

        if in_tags_list {
            if let Some(item) = trimmed.trim_start().strip_prefix("- ") {
                if let Some(tag) = clean_tag(item) {
                    parsed.frontmatter.tags.push(tag);
                }
                continue;
            }
            in_tags_list = false;
        }

        let Some((key, value)) = trimmed.split_once(':') else {
            continue;
        };
        let value = value.trim();

        match key.trim() {
            "pinned" => {
                parsed.frontmatter.pinned = value.eq_ignore_ascii_case("true");
            }
            "tags" => {
                if value.is_empty() {
                    in_tags_list = true;
                } else {
                    parsed.frontmatter.tags = parse_inline_tags(value);
                }
            }
            "title" => {
                parsed.frontmatter.title = clean_title(value);
            }
            _ => {}
        }
    }

    // Dedupe by value, not just adjacency -- the TypeScript twin uses a Set.
    let mut seen = HashSet::new();
    parsed
        .frontmatter
        .tags
        .retain(|tag| seen.insert(tag.clone()));

    parsed
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn no_frontmatter_returns_whole_body() {
        let parsed = parse("# hello\n");
        assert_eq!(parsed.body, "# hello\n");
        assert_eq!(parsed.frontmatter, Frontmatter::default());
    }

    #[test]
    fn parses_pinned_and_inline_tags() {
        let parsed = parse("---\npinned: true\ntags: [Errands, home]\n---\n- eggs\n");
        assert!(parsed.frontmatter.pinned);
        assert_eq!(parsed.frontmatter.tags, vec!["errands", "home"]);
        assert_eq!(parsed.body, "- eggs\n");
    }

    #[test]
    fn parses_block_list_tags() {
        let parsed = parse("---\ntags:\n  - \"work\"\n  - ideas\n---\nbody\n");
        assert_eq!(parsed.frontmatter.tags, vec!["work", "ideas"]);
        assert_eq!(parsed.body, "body\n");
    }

    #[test]
    fn ignores_unknown_keys_and_unclosed_blocks() {
        let parsed = parse("---\ncustom: thing\npinned: false\n---\nbody");
        assert!(!parsed.frontmatter.pinned);
        assert_eq!(parsed.body, "body");

        let unclosed = parse("---\npinned: true\nno close");
        assert!(!unclosed.frontmatter.pinned);
        assert_eq!(unclosed.body, "---\npinned: true\nno close");
    }

    #[test]
    fn dedupes_non_adjacent_tags() {
        let parsed = parse("---\ntags: [a, b, A]\n---\nbody");
        assert_eq!(parsed.frontmatter.tags, vec!["a", "b"]);
    }

    #[test]
    fn closes_on_a_delimiter_with_trailing_space() {
        let parsed = parse("---\npinned: true\n---  \nbody\n");
        assert!(parsed.frontmatter.pinned);
        assert_eq!(parsed.body, "body\n");

        let dots = parse("---\npinned: true\n...\nbody\n");
        assert!(dots.frontmatter.pinned);
        assert_eq!(dots.body, "body\n");
    }

    #[test]
    fn strips_separators_from_tags() {
        let parsed = parse("---\ntags:\n  - \"a,b\"\n---\nbody\n");
        assert_eq!(parsed.frontmatter.tags, vec!["ab"]);
    }

    #[test]
    fn parses_title_with_quotes_and_colons() {
        let plain = parse("---\ntitle: effect: a primer\n---\nbody\n");
        assert_eq!(
            plain.frontmatter.title,
            Some("effect: a primer".to_string())
        );

        let quoted = parse("---\ntitle: \"effect: a primer\"\n---\nbody\n");
        assert_eq!(
            quoted.frontmatter.title,
            Some("effect: a primer".to_string())
        );

        let single = parse("---\ntitle: 'my note'\n---\nbody\n");
        assert_eq!(single.frontmatter.title, Some("my note".to_string()));
    }

    #[test]
    fn treats_an_empty_title_as_absent() {
        let empty = parse("---\ntitle:\n---\nbody\n");
        assert_eq!(empty.frontmatter.title, None);

        let blank = parse("---\ntitle: \"\"\n---\nbody\n");
        assert_eq!(blank.frontmatter.title, None);

        let missing = parse("---\npinned: true\n---\nbody\n");
        assert_eq!(missing.frontmatter.title, None);
    }

    #[test]
    fn handles_crlf() {
        let parsed = parse("---\r\npinned: true\r\n---\r\nbody\r\n");
        assert!(parsed.frontmatter.pinned);
        assert_eq!(parsed.body, "body\r\n");
    }
}

File: src/core/frontmatter.ts
/**
 * Tolerant parser/serializer for the tiny frontmatter dialect notras cares
 * about. Only `pinned`, `tags` and `title` are interpreted; unknown keys are
 * preserved verbatim when rewriting, so externally-authored notes survive
 * round-trips. Kept in parity with the Rust parser in
 * `src-tauri/src/frontmatter.rs`.
 */

interface Frontmatter {
  pinned: boolean;
  tags: string[];
  /**
   * Read-only. notras resolves a title from this key but never authors it, so
   * it stays a foreign line and round-trips through `updateFrontmatter`.
   */
  title: string | undefined;
}

/** The keys `updateFrontmatter` writes. `title` is deliberately absent. */
interface FrontmatterPatch {
  pinned?: boolean;
  tags?: string[];
}

export interface ParsedNote {
  body: string;
  frontmatter: Frontmatter;
  /** Raw lines of the frontmatter block, without the `---` delimiters. */
  rawLines: string[];
}

const CLOSING_DELIMITERS = new Set(["---", "..."]);

function cleanTag(raw: string) {
  const tag = raw
    .trim()
    .replaceAll(/^["']+|["']+$/g, "")
    // Separators are dropped, never kept: a tag carrying one would serialize
    // into `tags: [a, b]` and re-parse as two tags.
    .replaceAll(/[,[\]]/g, "")
    .trim()
    .toLowerCase();

  return tag.length > 0 ? tag : undefined;
}

/**
 * A `title:` value with its quoting undone. An empty value counts as absent so
 * title resolution falls through to the next source.
 *
 * A single-quoted scalar has one escape, `''` for a literal quote, which is what
 * `serializeTitle` writes. A double-quoted one is only unwrapped: notras never
 * writes that form, and decoding YAML's backslash escapes would widen a dialect
 * `D6` keeps deliberately small. Kept in parity with `clean_title` in
 * `src-tauri/src/frontmatter.rs`.
 */
function cleanTitle(raw: string) {
  const value = raw.trim();
  const quote = value.length > 1 ? value.at(0) : undefined;
  const quoted =
    (quote === "'" || quote === '"') && value.endsWith(quote)
      ? value.slice(1, -1)
      : undefined;
  const title =
    quoted === undefined
      ? value
      : quote === "'"
        ? quoted.replaceAll("''", "'")
        : quoted;

  return title.length > 0 ? title : undefined;
}

/**
 * A title as a frontmatter value, quoted whenever plain YAML would misread it.
 *
 * Single quotes are the safe form for prose: the only escape inside one is `''`,
 * so a title carrying a backslash or a double quote needs no further handling,
 * where the double-quoted form would treat `\` as an escape introducer.
 */
function serializeTitle(title: string) {
  return /^[\p{L}\p{N}][\p{L}\p{N} \-._]*$/u.test(title)
    ? title
    : `'${title.replaceAll("'", "''")}'`;
}

function parseInlineTags(value: string) {
  return value
    .trim()
    .replace(/^\[+/, "")
    .replace(/\]+$/, "")
    .split(",")
    .map(cleanTag)
    .filter((tag) => tag !== undefined);
}

export function parseNote(content: string): ParsedNote {
  const fallback: ParsedNote = {
    body: content,
    frontmatter: { pinned: false, tags: [], title: undefined },
    rawLines: [],
  };

  if (!(content.startsWith("---\n") || content.startsWith("---\r\n"))) {
    return fallback;
  }

  const lines = content.split("\n");
  const closeIndex = lines.findIndex(
    (line, index) => index > 0 && CLOSING_DELIMITERS.has(line.trimEnd())
  );

  if (closeIndex === -1) {
    return fallback;
  }

  // CRLF files leave a trailing \r on every split line; strip it here so
  // `composeNote`'s \n joins cannot emit a block with mixed line endings.
  const rawLines = lines
    .slice(1, closeIndex)
    .map((line) => line.replace(/\r$/, ""));
  const body = lines.slice(closeIndex + 1).join("\n");
  const frontmatter: Frontmatter = {
    pinned: false,
    tags: [],
    title: undefined,
  };

  let inTagsList = false;

  for (const line of rawLines) {
    const trimmed = line.trimEnd();

    if (inTagsList) {
      const item = trimmed.trimStart();

      if (item.startsWith("- ")) {
        const tag = cleanTag(item.slice(2));

        if (tag !== undefined) {
          frontmatter.tags.push(tag);
        }

        continue;
      }

      inTagsList = false;
    }

    const separator = trimmed.indexOf(":");

    if (separator === -1) {
      continue;
    }

    const key = trimmed.slice(0, separator).trim();
    const value = trimmed.slice(separator + 1).trim();

    switch (key) {
      case "pinned": {
        frontmatter.pinned = value.toLowerCase() === "true";

        break;
      }
      case "tags": {
        if (value === "") {
          inTagsList = true;
        } else {
          frontmatter.tags = parseInlineTags(value);
        }

        break;
      }
      case "title": {
        frontmatter.title = cleanTitle(value);

        break;
      }
      // No default
    }
  }

  frontmatter.tags = [...new Set(frontmatter.tags)];

  return { body, frontmatter, rawLines };
}

/** Drops the `pinned`/`tags` lines (and tag list items) from a raw block. */
function withoutOwnKeys(rawLines: string[]) {
  let inTagsList = false;

  return rawLines.filter((line) => {
    const trimmed = line.trimEnd();

    if (inTagsList) {
      if (trimmed.trimStart().startsWith("- ")) {
        return false;
      }

      inTagsList = false;
    }

    const separator = trimmed.indexOf(":");

    if (separator === -1) {
      return true;
    }

    const key = trimmed.slice(0, separator).trim();

    if (key === "pinned") {
      return false;
    }

    if (key === "tags") {
      inTagsList = trimmed.slice(separator + 1).trim() === "";

      return false;
    }

    return true;
  });
}

/**
 * Rewrite an existing `title:` line in place, keeping its position in the block
 * and its indentation, or return `rawLines` unchanged when there is no such
 * key. notras updates a title key but never introduces one.
 *
 * The key is matched the way `parseNote` reads it, and the last match wins for
 * the same reason: a duplicated key is invalid YAML, and the two must agree
 * about which one is live. `serializeTitle` decides the quoting, so a title
 * carrying a colon, a quote, or a backslash stays valid YAML for the other tools
 * reading the file.
 */
export function retitleFrontmatter(rawLines: string[], title: string) {
  const index = rawLines.findLastIndex((line) => {
    const trimmed = line.trimEnd();
    const separator = trimmed.indexOf(":");

    return separator !== -1 && trimmed.slice(0, separator).trim() === "title";
  });

  if (index === -1) {
    return rawLines;
  }

  const existing = rawLines[index] ?? "";
  const indent = existing.slice(
    0,
    existing.length - existing.trimStart().length
  );

  return rawLines.with(index, `${indent}title: ${serializeTitle(title)}`);
}

/** Reassemble a full note file from raw frontmatter lines and a body. */
export function composeNote(rawLines: string[], body: string) {
  if (rawLines.length === 0) {
    return body;
  }

  return `---\n${rawLines.join("\n")}\n---\n${body}`;
}

/**
 * Rewrite a note's frontmatter with new `pinned`/`tags` values, preserving
 * every unknown key. Default values (`pinned: false`, no tags) are omitted;
 * a block that ends up empty is removed entirely.
 *
 * `title` is not writable and is not one of `withoutOwnKeys`'s own keys, so a
 * note carrying one keeps it verbatim through a pin or tag toggle.
 */
export function updateFrontmatter(
  content: string,
  patch: FrontmatterPatch
): string {
  const parsed = parseNote(content);
  const next = {
    pinned: patch.pinned ?? parsed.frontmatter.pinned,
    tags: patch.tags ?? parsed.frontmatter.tags,
  };

  const foreignLines = withoutOwnKeys(parsed.rawLines);

  const ownLines: string[] = [];

  if (next.pinned) {
    ownLines.push("pinned: true");
  }

  // Clean on the way out too, so a tag can never carry a separator into the
  // inline form and come back as two tags.
  const tags = [
    ...new Set(next.tags.map(cleanTag).filter((tag) => tag !== undefined)),
  ];

  if (tags.length > 0) {
    ownLines.push(`tags: [${tags.join(", ")}]`);
  }

  const blockLines = [...ownLines, ...foreignLines];

  if (blockLines.length === 0) {
    return parsed.body;
  }

  return `---\n${blockLines.join("\n")}\n---\n${parsed.body}`;
}
